Adds Module 6: Testing Code (#125)
rsokl/Learning_Python: Python/Module6_Testing/Hypothesis.md

    name: conda-env-.conda-jupy-py
---

<!-- #raw raw_mimetype="text/restructuredtext" -->
.. meta::
   :description: Topic: Writing tests for your code, Difficulty: Easy, Category: Section
   :keywords: test, automated, pytest, parametrize, fixture, suite, decorator, clean directory  
<!-- #endraw -->

<!-- #region -->
# Introduction to Testing with Hypothesis

It is often the case that the process of *describing our data* is by far the heaviest burden that we must bear when writing tests. This process of assessing "what variety of values should I test?", "have I thought of all the important edge-cases?", and "how much is 'enough'?" will crop up with nearly every test that we write.
Indeed, these are questions that you may have been asking yourself when writing `test_count_vowels_basic` and `test_merge_max_mappings` in the previous sections of this module.

[Hypothesis](https://hypothesis.readthedocs.io/) is a powerful Python library that empowers us to write a _description_ (specification, to be more precise) of the data that we want to use to exercise our test.
It will then *generate* test cases that satisfy this description and will run our test on these cases.

> Ultimately this an extremely powerful tool for enabling us to write high-quality automated tests for our code.

Reviewer: ```suggestion
Ultimately, this an extremely powerful tool for enabling us to write high-quality automated tests for our code.
```

# `size` can be any non-negative integer
@given(size=st.integers(min_value=0))
def test_range_length(size):
    assert len(range(size)) == size
```

which places no formal upper bound on the integers that Hypothesis will generate.
However, this test immediately found an issue (I hesitate to call it an outright bug):

```python
Falsifying example: test_range_length(
    size=9223372036854775808,
)

----> 3     assert len(range(size)) == size

OverflowError: Python int too large to convert to C ssize_t
```

This reveals that the CPython implementation of the built-in `len` function is such that it can only handle non-negative integers smaller than $2^{63}$ (i.e. it will only allocate 64 bits to represent a signed integer - one bit is used to store the sign of the number).
Hypothesis revealed this by generating the failing test case `size=9223372036854775808`, which is exactly $2^{63}$.
I did not want this error to distract from what is otherwise merely a simple example, but it is very important to point out.

Hypothesis has a knack for catching these sorts of unexpected edge cases.
Now we know that `len(range(size)) == size` _does not_ hold for "arbitrary" non-negative integers!

> (This overflow behavior is now documented in the [CPython source code](https://github.com/python/cpython) because it was discovered while writing this material 😄).

Reviewer 2: It's documented in the docs too these days!

I'd avoid giving the impression that it was unknown before; the core devs knew but just hadn't documented the implementation detail.

</div>
<!-- #endregion -->

<!-- #region -->
## The `given` Decorator

Hypothesis' [given decorator](https://hypothesis.readthedocs.io/en/latest/details.html#the-gory-details-of-given-parameters) is responsible for:

 - drawing values from Hypothesis' so-called "strategies" for describing input data for our test function
 - running the test function many times (up to 100 times, by default) given different input values drawn from the strategy

>  - "shrinking" the drawn inputs to identify simple fail cases: if an error is raised by the test function during one of the many execution, the `given` decorator will attempt to "shrink" (i.e. simplify) the inputs that produce that same error before reporting them to the user

Reviewer: ```suggestion
 - "shrinking" the drawn inputs to identify simple fail cases: if an error is raised by the test function during one of the many executions, the `given` decorator will attempt to "shrink" (i.e. simplify) the inputs that produce that same error before reporting them to the user
```

)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-29-ea20353bbef5> in test_demonstrating_the_given_decorator(x, y)
     10     # `y` can be any value in [20, 30]
     11     # this is a bad assertion: it should fail!
---> 12     assert 20 <= y <= 25

AssertionError: 

```

(Note: some of the "Traceback" error message has been removed to improve legibility)

See that the error message here indicates that Hypothesis identified a "falsifying example", or a set of input values, `x=0` and `y=26`, which caused our test function to raise an error. The proceeding "Traceback" message indicates that it is indeed the second assertion statement that is responsible for raising the error.

### Shrinking: Simplifying Falsifying Inputs

The `given` decorator strives to report the "simplest" set of input values that produce a given error.
It does this through the process of "shrinking".
Each of Hypothesis' strategies has its own prescribed shrinking behavior.
For the `integers` strategy, this means identifying the integer closest to 0 that produces the error at hand.

> For instance, `x=12` and `y=29` may have been the first drawn values to trigger the assertion error.

Reviewer: not really that relevant to the discussion right here but are you planning to talk at all about how hypothesis actually draws data?

Author: @Zac-HD do you think there is anything sufficiently concise and informative to say towards this end, given the basic nature of this discussion?

Reviewer 2: Nope... I'd just leave it totally unspoken and they can go to Hypothesis documentation if curious.

Which is almost as silent on the topic, just "it's reproducible, arbitrary, and don't try to guess the distribution".  Anything further is internals and subject to change without notice.

Hypothesis provides a useful mechanism for developing an intuition for the data produced by a strategy: a strategy, once initialized, has a `.example()` method that will randomly draw a representative value from the strategy. For example:

```python
# demonstrating usage of `<strategy>.example()`
>>> st.integers(-1, 1).example()
-1

>>> st.integers(-1, 1).example()
1
```

**Note: the** `.example()` **mechanism is only meant to be used for pedagogical purposes. You should never use this in your test suite**
because (among other reasons) `.example()` biases towards smaller and simpler examples than `@given`, and lacks the features to ensure any test failures are reproducible.

We will be leveraging the `.example()` method throughout the rest of this section to help provide an intuition for the data that Hypothesis' various strategies generate.
<!-- #endregion -->

<!-- #region -->
### Exploring Strategies

There are a number critical Hypothesis strategies for us to become familiar with. It is worthwhile to peruse through all of Hypothesis' [core strategies](https://hypothesis.readthedocs.io/en/latest/data.html#core-strategies), but we will take time to highlight a few here.

> #### `st.booleans ()`

Reviewer 2: ```suggestion
#### `st.booleans()`
```

[st.booleans()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.booleans) generates either `True` or `False`. This strategy will shrink towards `False`

```python
>>> st.booleans().example()
False
```
<!-- #endregion -->

<!-- #region -->

> #### `st.lists ()`

Reviewer 2: ```suggestion
#### `st.lists()`
```

 - if we want the elements to be unique
 - a mechanism for defining "uniqueness"

For example, the following strategy describes lists whose length varies from 2 to 10, and whose entries are integers on the domain $[-10, 20]$:

```python
>>> st.lists(st.integers(-10, 20), min_size=2, max_size=10).example()
[-10, 0, 5]
```

**`st.lists(...)` is the strategy of choice anytime we want to generate sequences of varying lengths with elements that are, themselves, described by strategies**.
<!-- #endregion -->

<!-- #region -->
#### `st.floats()`

[st.floats](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.floats) is a powerful strategy that generates all variety of floats, including `math.inf` and `math.nan`. You can also specify:
 - whether `math.inf` and `math.nan`, respectively, should be included in the data description
 - bounds (either inclusive or exclusive) on the floats being generated; this will naturally preclude `math.nan` from being generated
 - the "width" of the floats; e.g. if you want to generate 16-bit or 32-bit floats vs 64-bit
   (while Python's `float` is (usually) 64-bit, `width=32` ensures that the generated values can

>    always be losslessly represented in 32 bits.  This is mostly useful for Numpy arrays.)

Reviewer: ```suggestion
   always be losslessly represented in 32 bits.  This is mostly useful for NumPy arrays.)
```

For example, the following strategy 64-bit floats that reside in the domain $[-100, 1]$:

```python
>>> st.floats(-100, 1).example()
0.3670816313319896
```
<!-- #endregion -->

<!-- #region -->
#### `st.tuples()`

The [st.tuples](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.tuples) strategy accepts $N$ Hypothesis strategies, and will generate length-$N$ tuples whose elements are drawn from the respective strategies that were specified as inputs.

For example, the following strategy will generate length-3 tuples whose entries are: integers, booleans, and floats:

```python
>>> st.tuples(st.integers(), st.booleans(), st.floats()).example()
(4628907038081558014, False, -inf)
```
<!-- #endregion -->

<!-- #region -->
#### `st.text()`

> The [st.text](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.text) accepts an "alphabet" – a collection of string-characters – from which it will construct strings of varying lengths, whose bounds can be specified by the user.

Reviewer 2: The alphabet argument must be a collection of length-one strings (note that a string is such a collection), or a strategy for generating such values (eg st.characters()).

> For example, the following strategy will strings of lowercase vowels from length 2 to length 10:

Reviewer 2: ```suggestion
For example, the following strategy will generate strings of lowercase vowels from length 2 to length 10:
```

```python
>>> st.text("aeiouy", min_size=2, max_size=10).example()
'oouoyoye'
```
<!-- #endregion -->

<!-- #region -->
#### `st.just()`

[st.just](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.just) is a strategy that "just" returns the value that you fed it. This is a convenient strategy that helps us to avoid having to manipulate our data before using it.

> Write suppose that we want a strategy that describes the shape of an array (i.e. a tuple of integers) that contains 1-20 two-dimensional vectors. E.g. `(5, 2)` is the shape of the array containing five two-dimensional vectors. We can leverage `st.just`, in conjunction with `st.integers` and `st.tuples`, towards this end:

Reviewer: ```suggestion
Suppose that we want a strategy that describes the shape of an array (i.e. a tuple of integers) that contains 1-20 two-dimensional vectors. E.g. `(5, 2)` is the shape of the array containing five two-dimensional vectors. We can leverage `st.just`, in conjunction with `st.integers` and `st.tuples`, towards this end:
```

```python
>>> st.tuples(st.integers(1, 20), st.just(2)).example()
(7, 2)
```
<!-- #endregion -->

<!-- #region -->
#### `st.one_of()`

> The [st.one_of](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.one_of) allows us to specify a collection of strategies and any given datum will be drawn from "one of" them. E.g.

Reviewer: ```suggestion
The [st.one_of](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.one_of) strategy allows us to specify a collection of strategies and any given datum will be drawn from "one of" them. For example:
```

```python
# demonstrating st.one_of()
st.one_of(st.integers(), st.lists(st.integers()))
```

> will draw values that are *either* integers or list of integers:

Reviewer: ```suggestion
will draw values that are *either* integers or lists of integers:
```

```python
>>> st.one_of(st.integers(), st.lists(st.integers())).example()
144

>>> st.one_of(st.integers(), st.lists(st.integers())).example()
[0, -22]
```

The "pipe" operator, `|` can be used between strategies, to chain `st.one_of` calls:

```python

> # Using the pipe operation, | , in place of `st.one_of`

Reviewer: ```suggestion
# Using the pipe operator, | , in place of `st.one_of`
```

# This strategy generates integers or floats
# or lists that contain just the word "hello"

>>> (st.integers() | st.floats() | st.lists(st.just("hello"))).example()
['hello', 'hello']

>>> (st.integers() | st.floats() | st.lists(st.just("hello"))).example()
0
```
<!-- #endregion -->

<!-- #region -->

> #### `st.sampled_from`

Reviewer 2: ```suggestion
#### `st.sampled_from()`
```

> [st.sampled_from](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.sampled_from) accepts a collection of objects (anything that has a length and supports integer-based indexing is a collection; e.g. lists, tuples, strings, and numpy arrays). The strategy will return values that are sampled from this collections.

Reviewer: ```suggestion
[st.sampled_from](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.sampled_from) accepts a collection of objects (anything that has a length and supports integer-based indexing is a collection; e.g. lists, tuples, strings, and NumPy arrays). The strategy will return values that are sampled from this collection.
```

> For example, the following strategy will sample from a list either `0`, `"a"`, or `(2, 2)`

Reviewer: ```suggestion
For example, the following strategy will sample a value `0`, `"a"`, or `(2, 2)` from a list:
```

something like this reads a little better to me

```python
>>> st.sampled_from([0, "a", (2, 2)]).example()
'a'
```
<!-- #endregion -->

<div class="alert alert-info"> 

**Reading Comprehension: Exploring other Core Strategies**

Review the [rest of Hypothesis' core strategies](https://hypothesis.readthedocs.io/en/latest/data.html#core-strategies).
Write down a strategy, and print out a representative example, that describes the the data according to each of the following conditions:

>    1. Dictionaries of arbitrary size whose keys are positive-values integers and whose values are `True` or `False.

Reviewer: ```suggestion
   1. Dictionaries of arbitrary size whose keys are positive-valued integers and whose values are `True` or `False.
```

   2. Length-4 strings whose elements are only lowercase vowels
   3. Permutations of the list `[1, 2, 3, 4]`

</div>

<!-- #region -->
<div class="alert alert-info">

**Reading Comprehension: Improving our tests using Hypothesis**

We will be writing improved tests for the basic functions – `count_vowels` and `merge_max_mappings` – by leveraging Hypothesis.
This reading comprehension question will require more substantial work than usual.
That being said, the experience that we will gain from this will be well worth the work.
Keep in mind that solutions are included at the end of this page, and that these can provide guidance if we get stuck.

Part 1: Testing correctness by construction

> Write a hypothesis-driven test for the `count_vowels`; include this test in `test/test_basic_functions`.

Reviewer: ```suggestion
Write a Hypothesis-driven test for the `count_vowels`; include this test in `tests/test_basic_functions`.
```

for consistency

> This is a test function where we can explicit construct a string in parts: its non-vowel characters, non-y vowels, and y-vowels.

Reviewer: ```suggestion
This is a test function where we can explicitly construct a string in parts: its non-vowel characters, non-y vowels, and y-vowels.
```

And thus, by constructing a string with a known number of vowel and non-vowel characters, we can know what the output of `count_vowels` *should* be for that input, and we can thus test for correctness in this way.
We will want to read about the [st.text()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.text) strategy to construct the different parts of the string.
The standard library's built-in `string` module provides a string of all printable characters (`string.printable`).

> We should ask ourselves: How general are input strings that we are constructing? Are there regular patterns in the strings that might prevent our test from identifying edge case bugs in `count_vowels`?

Reviewer: ```suggestion
We should ask ourselves: how general are input strings that we are constructing? Are there regular patterns in the strings that might prevent our test from identifying edge case bugs in `count_vowels`?
```

Part 2: Property-based testing

> Write a hypothesis-driven test for `merge_max_mappings` ; include this test in `test/test_basic_functions`.

Reviewer: ```suggestion
Write a Hypothesis-driven test for `merge_max_mappings` ; include this test in `tests/test_basic_functions`.
```

Here, we can't simply contrive the inputs to `merge_max_mappings` in a general way and know what its output should be – we would have to re-implement the function to do that.
Instead, we should *test the expected properties* of the merged dictionary.
For example, one such property is that the merged dictionary should only contain maximum values.
Another property would be that all of the keys among the input dictionaries should be present in the merged dictionary.
Take some time to think of other such properties that we should test for.
Ultimately we want to arrive at a comprehensive set of properties to test for such that we can be confident that our merged dictionary is correct.

We will want to use [st.dictionaries()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.dictionaries) to describe the inputs to `merge_max_mappings`.
Although dictionary keys can be any hashable object, suffice it to use both integers and text for the keys, and integers for the dictionary values in this test, for simplicity's sake.

> **We must remember to temporarily mutate our original functions to verify that these tests can actually catch bugs!**

Reviewer 2: Might want a link to the idea of mutation testing here?

Author: I tell the reader about mutation testing and link them to the cosmic ray library earlier. Here I just want them to do some simple by-hand mutating

# valued integers. This is achieved by ensuring that any integer
# drawn by `st.integers()` is "mapped" to an even value
# via the function x -> 2 * x
even_integers = st.integers().map(make_even)
```

```python
>>> even_integers.example()
-15414
```
<!-- #endregion -->

<!-- #region -->
#### A Brief Aside: Lambda Expressions

Python has a syntax, which we have yet to discuss, that permits us to conveniently define functions "on the fly".
A "lambda expression" is a syntax for defining a simple one-line function, making that function available for use in-place.
The key here is that, whereas standard functions first must be formally defined before they can be referenced in code, *a lambda expression can be written wherever a function is expected*.

For example, we can simplify the above mapping example by defining our mapping-function _within_ the `.map()` method:

```python
# Using a lambda expression to define a function
# "on the fly"
even_integers = st.integers().map(lambda x: 2 * x)
```

```python
>>> even_integers.example()
220
```

In general, the syntax for defining a lambda expression is:

```
lambda <comma-separated variable names>: <expression using variables>
```

> Note that lambda expressions are restricted compared to typical function definitions: they do not permit default values or keyword arguments in their signatures, and a lambda's "body" must fit on one line.

Reviewer 2: Lambdas *do* permit default values and everything else you could put in a normal function signature; the restriction is that the function body must consist of a single expression - but that can span multiple physical lines!

Author: 😮 I'm ashamed that I didn't know this!

---

<!-- #raw raw_mimetype="text/restructuredtext" -->
.. meta::
   :description: Topic: Writing tests for your code, Difficulty: Easy, Category: Section
   :keywords: test, automated, pytest, parametrize, fixture, suite, decorator, clean directory  
<!-- #endraw -->

<!-- #region -->
# Describing Data with Hypothesis

It is often the case that the process of *describing our data* is by far the heaviest burden that we must bear when writing tests. This process of assessing "what variety of values should I test?", "have I thought of all the important edge-cases?", and "how much is 'enough'?" will crop up with nearly every test that we write.
Indeed, these are questions that you may have been asking yourself when writing `test_count_vowels_basic` and `test_merge_max_mappings` in the previous sections of this module.

[Hypothesis](https://hypothesis.readthedocs.io/) is a powerful Python library that empowers us to write a _description_ (specification, to be more precise) of the data that we want to use to exercise our test.
It will then *generate* test cases that satisfy this description and will run our test on these cases.

Let's look at a simple example of Hypothesis in action.

> In the preceding section, we learned to use pytest's parameterization mechanism to test properties of code over a set of values.

Author: ![image](https://user-images.githubusercontent.com/29104956/72206269-a39e0f80-3459-11ea-861d-d4cc66eeafe1.png)

   :keywords: test, automated, pytest, parametrize, fixture, suite, decorator, clean directory  
<!-- #endraw -->

<!-- #region -->
# Introduction to Testing with Hypothesis

It is often the case that the process of *describing our data* is by far the heaviest burden that we must bear when writing tests. This process of assessing "what variety of values should I test?", "have I thought of all the important edge-cases?", and "how much is 'enough'?" will crop up with nearly every test that we write.
Indeed, these are questions that you may have been asking yourself when writing `test_count_vowels_basic` and `test_merge_max_mappings` in the previous sections of this module.

[Hypothesis](https://hypothesis.readthedocs.io/) is a powerful Python library that empowers us to write a _description_ (specification, to be more precise) of the data that we want to use to exercise our test.
It will then *generate* test cases that satisfy this description and will run our test on these cases.
Ultimately this an extremely powerful tool for enabling us to write high-quality automated tests for our code.

Hypothesis can be installed via conda:

```shell

> conda install -c conda-forge hypothesis

Reviewer 2: Hypothesis is [available from the default channel](https://anaconda.org/anaconda/hypothesis) as well as [conda-forge](https://anaconda.org/conda-forge/hypothesis).  The default channel is usually a week to a month behind, but since most `conda` users don't update regularly that probably doesn't matter much...  I'd just show the defaults install here.

# `size` can be any non-negative integer
@given(size=st.integers(min_value=0))
def test_range_length(size):
    assert len(range(size)) == size
```

which places no formal upper bound on the integers that Hypothesis will generate.
However, this test immediately found an issue (I hesitate to call it an outright bug):

```python
Falsifying example: test_range_length(
    size=9223372036854775808,
)

----> 3     assert len(range(size)) == size

OverflowError: Python int too large to convert to C ssize_t
```

This reveals that the CPython implementation of the built-in `len` function is such that it can only handle non-negative integers smaller than $2^{63}$ (i.e. it will only allocate 64 bits to represent a signed integer - one bit is used to store the sign of the number).
Hypothesis revealed this by generating the failing test case `size=9223372036854775808`, which is exactly $2^{63}$.
I did not want this error to distract from what is otherwise merely a simple example, but it is very important to point out.

Hypothesis has a knack for catching these sorts of unexpected edge cases.
Now we know that `len(range(size)) == size` _does not_ hold for "arbitrary" non-negative integers!

> (I wonder how many of the Python core developers know about this 😄).

Reviewer 2: They know!  After we worked out what was happening I wrote https://github.com/python/cpython/pull/17934 so it's even documented now :joy:

<!-- #region -->
### Exploring Strategies

There are a number critical Hypothesis strategies for us to become familiar with. It is worthwhile to peruse through all of Hypothesis' [core strategies](https://hypothesis.readthedocs.io/en/latest/data.html#core-strategies), but we will take time to highlight a few here.

#### `st.booleans()`

[st.booleans()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.booleans) generates either `True` or `False`. This strategy will shrink towards `False`

```python
>>> st.booleans().example()
False
```
<!-- #endregion -->

<!-- #region -->

#### `st.lists()`

[st.lists()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.lists) accepts *another* strategy, which describes the elements of the lists being generated. You can also specify:
 - bounds on the length of the list
 - if we want the elements to be unique
 - a mechanism for defining "uniqueness"

For example, the following strategy describes lists whose length varies from 2 to 10, and whose entries are integers on the domain $[-10, 20]$:

```python
>>> st.lists(st.integers(-10, 20), min_size=2, max_size=10).example()
[-10, 0, 5]
```

> **st.lists(...) is our go-to anytime we want to create a strategy that generates sequences of varying lengths with elements that are, themselves, described by strategies**.

Reviewer 2: ```suggestion
**`st.lists(...)` is our go-to anytime we want to create a strategy that generates sequences of varying lengths with elements that are, themselves, described by another strategy.**
```

<!-- #endregion -->

<!-- #region -->
#### `st.floats()`

[st.floats()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.floats) is a powerful strategy that generates all variety of floats, including `math.inf` and `math.nan`. You can also specify:
 - whether `math.inf` and `math.nan`, respectively, should be included in the data description
 - bounds (either inclusive or exclusive) on the floats being generated; this will naturally preclude `math.nan` from being generated
 - the "width" of the floats; e.g. if you want to generate 16-bit or 32-bit floats vs 64-bit
   (while Python's `float` is (usually) 64-bit, `width=32` ensures that the generated values can
   always be losslessly represented in 32 bits.  This is mostly useful for NumPy arrays.)

For example, the following strategy 64-bit floats that reside in the domain $[-100, 1]$:

```python
>>> st.floats(-100, 1).example()
0.3670816313319896
```
<!-- #endregion -->

<!-- #region -->
#### `st.tuples()`

> The [st.tuples](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.tuples) strategy accepts $N$ Hypothesis strategies, and will generate length-$N$ tuples whose elements are drawn from the respective strategies that were specified as inputs.

Reviewer 2: ```suggestion
The [st.tuples()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.tuples) strategy accepts $N$ Hypothesis strategies, and will generate length-$N$ tuples whose elements are drawn from the respective strategies that were specified as inputs.
```

For example, the following strategy will generate length-3 tuples whose entries are: integers, booleans, and floats:

```python
>>> st.tuples(st.integers(), st.booleans(), st.floats()).example()
(4628907038081558014, False, -inf)
```
<!-- #endregion -->

<!-- #region -->
#### `st.text()`

The [st.text()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.text) strategy accepts an "alphabet" – a collection of length-one strings or a strategy for generating such values (such as `st.characters()`) – from which it will construct strings of varying lengths, whose bounds can be specified by the user.

For example, the following strategy will generate strings of lowercase vowels from length 2 to length 10:

```python
>>> st.text("aeiouy", min_size=2, max_size=10).example()
'oouoyoye'
```
<!-- #endregion -->

<!-- #region -->
#### `st.just()`

[st.just()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.just) is a strategy that "just" returns the value that you fed it. This is a convenient strategy that helps us to avoid having to manipulate our data before using it.

> Suppose that we want a strategy that describes the shape of an array (i.e. a tuple of integers) that contains 1-20 two-dimensional vectors. E.g. `(5, 2)` is the shape of the array containing five two-dimensional vectors. We can leverage `st.just`, in conjunction with `st.integers` and `st.tuples`, towards this end:

Reviewer 2: ```suggestion
Suppose that we want a strategy that describes the shape of an array (i.e. a tuple of integers) that contains 1-20 two-dimensional vectors. E.g. `(5, 2)` is the shape of the array containing five two-dimensional vectors. We can leverage `st.just()`, in conjunction with `st.integers()` and `st.tuples()`, towards this end:
```

```python
>>> st.tuples(st.integers(1, 20), st.just(2)).example()
(7, 2)
```
<!-- #endregion -->

<!-- #region -->
#### `st.one_of()`

> The [st.one_of](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.one_of) strategy allows us to specify a collection of strategies and any given datum will be drawn from "one of" them. For example:

Reviewer 2: ```suggestion
The [st.one_of()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.one_of) strategy allows us to specify a collection of strategies and any given datum will be drawn from "one of" them. For example:
```

```python
# demonstrating st.one_of()
st.one_of(st.integers(), st.lists(st.integers()))
```

will draw values that are *either* integers or lists of integers:

```python
>>> st.one_of(st.integers(), st.lists(st.integers())).example()
144

>>> st.one_of(st.integers(), st.lists(st.integers())).example()
[0, -22]
```

The "pipe" operator, `|` can be used between strategies, to chain `st.one_of` calls:

```python
# Using the pipe operator, | , in place of `st.one_of`
# This strategy generates integers or floats
# or lists that contain just the word "hello"

>>> (st.integers() | st.floats() | st.lists(st.just("hello"))).example()
['hello', 'hello']

>>> (st.integers() | st.floats() | st.lists(st.just("hello"))).example()
0
```
<!-- #endregion -->

<!-- #region -->
#### `st.sampled_from()`

> [st.sampled_from](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.sampled_from) accepts a collection of objects (anything that has a length and supports integer-based indexing is a collection; e.g. lists, tuples, strings, and NumPy arrays) and returns a strategy that are randomly samples values from this collection.

Reviewer 2: ```suggestion
[st.sampled_from()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.sampled_from) accepts a sequence of objects (anything that has a length and supports integer-based indexing is a sequence; e.g. lists, tuples, strings, and NumPy arrays) and returns a strategy that are randomly samples values from this sequence.
```

For example, the following strategy will sample a value `0`, `"a"`, or `(2, 2)` from a list:

```python
>>> st.sampled_from([0, "a", (2, 2)]).example()
'a'
```
<!-- #endregion -->

<div class="alert alert-info"> 

**Reading Comprehension: Exploring other Core Strategies**

Review the [rest of Hypothesis' core strategies](https://hypothesis.readthedocs.io/en/latest/data.html#core-strategies).
Write down a strategy, and print out a representative example, that describes the the data according to each of the following conditions:

>    1. Dictionaries of arbitrary size whose keys are positive-valued integers and whose values are `True` or `False.

Reviewer 2: ```suggestion
   1. Dictionaries of arbitrary size whose keys are positive-valued integers and whose values are `True` or `False`.
```

# valued integers. This is achieved by ensuring that any integer
# drawn by `st.integers()` is "mapped" to an even value
# via the function x -> 2 * x
even_integers = st.integers().map(make_even)
```

```python
>>> even_integers.example()
-15414
```
<!-- #endregion -->

<!-- #region -->
#### A Brief Aside: Lambda Expressions

Python has a syntax, which we have yet to discuss, that permits us to conveniently define functions "on the fly".
A "lambda expression" is a syntax for defining a simple one-line function, making that function available for use in-place.
The key here is that, whereas standard functions first must be formally defined before they can be referenced in code, *a lambda expression can be written wherever a function is expected*.

For example, we can simplify the above mapping example by defining our mapping-function _within_ the `.map()` method:

```python
# Using a lambda expression to define a function
# "on the fly"
even_integers = st.integers().map(lambda x: 2 * x)
```

```python
>>> even_integers.example()
220
```

In general, the syntax for defining a lambda expression is:

```
lambda <comma-separated variable names>: <expression using variables>
```

> Note that lambda expressions are restricted compared to typical function definitions: their body must consist only of a single Python expression, and the lambda function returns whatever that expression returns.

Reviewer 2: ```suggestion
Note that lambda expressions are restricted compared to typical function definitions: their body must consist only of a single Python expression, and the lambda function returns whatever value that expression evaluates to.
```

it should simply test that either `"cat"` or `"dog"` was indeed produced by the strategy.
Run the test.

</div>

<!-- #region -->
### The `.filter` method

Hypothesis strategies can also [have their data filtered](https://hypothesis.readthedocs.io/en/latest/data.html#filtering) via the `.filter` method. 
`.filter` takes a function (or any "callable") that accepts as input the data generated by the strategy, and returns:

 - `True` if the data should pass through the filter
 - `False` if the data should be rejected by the filter

Consider, for instance, that you want to generate all integers other than `0`.
You can write the filtered strategy:

```python
# Demonstrating the `.filter()` method
non_zero_integers = st.integers().filter(lambda x: x != 0)
```

The `.filter` method is not magic – it is not able to "just know" how to avoid generating all barred values. 
A strategy that filters our too much data will prompt Hypothesis to raise an error.
For example, let's try to filter `st.integers()` so that it only produces values on $[10, 20]$.

```python
# Using `.filter()` to filter out a large proportion of generated
# values will result in an error

> @given(st.integers().filter(lambda x: 10 <= x <= 20))

Reviewer 2: Unfortunately (?), this example *does now work* thanks to our filter-rewriting logic for intervals!  I suggest using "multiples of 17" or similar.

<!-- #endregion -->

<!-- #region -->
**Fixing the Failing Test: Solution**

Update the body of `test_demonstrating_the_given_decorator` so that it no longer fails.

> We simply need to fix the second assertion statement, specifying the bounds on `y`, so that it agrees with what is being drawn from the `integers` strategy.

```python
@given(x=st.integers(0, 10), y=st.integers(20, 30))
def test_demonstrating_the_given_decorator(x, y):
    assert 0 <= x <= 10
    assert 20 <= y <= 30
```

Run the fixed test function. How many times is the test function actually be executed when you run it?

> The `given` decorator, by default, will draw 100 sets of example values from the strategies that are passed to it and will thus execute the decorated test function 100 times.

```python
# no output (the function returns `None`) means that the test passed
>>> test_demonstrating_the_given_decorator()
```
<!-- #endregion -->

<!-- #region -->
**Exploring other Core Strategies: Solution**

> Dictionaries of arbitrary size whose keys are positive-values integers and whose values are `True` or `False.

Reviewer 2: ```suggestion
Dictionaries of arbitrary size whose keys are positive-values integers and whose values are `True` or `False`.
```

<!-- #region -->
```python
from string import printable
from random import shuffle

import hypothesis.strategies as st
from hypothesis import given, note

# a list of all printable non-vowel characters
_not_vowels = "".join([l for l in printable if l.lower() not in set("aeiouy")])

@given(
    not_vowels=st.text(alphabet=_not_vowels),
    vowels_but_not_ys=st.text(alphabet="aeiouAEIOU"),
    ys=st.text(alphabet="yY"),
)
def test_count_vowels_hypothesis(not_vowels, vowels_but_not_ys, ys):
    """
    Constructs an input string with a known number of:
       - non-vowel characters
       - non-y vowel characters
       - y characters
    
    and thus, by constructions, we can test that the output
    of `count_vowels` agrees with the known number of vowels
    """
    # list of characters
    letters = list(not_vowels) + list(vowels_but_not_ys) + list(ys)
    
    # We need to shuffle the ordering of our characters so that
    # our input string isn't unnaturally patterned; e.g. always 
    # have its vowels at the end
    shuffle(letters)

>     in_string = "".join(letters)

Reviewer 2: Trouble!  Hypothesis will seed the random-number generator, so you shouldn't show using `random` functions in PBTs.

Or, preferably, use the `st.random()` strategy to provide a `Random` instance that you use to do the shuffle, and explain why that's a good idea.

Using the `.map()` method, construct a Hypothesis strategy that produces a sorted list of integers.
Generate some examples from your strategy and check that they are sorted (we may have to generate quite a few examples to see a diverse set of values)

<!-- #region -->
```python
# Note that the built-in `sorted` function can be supplied 
# directly to the `.map()` method - there is no need to define
# a function or use a lambda expression here
sorted_list_of_ints = st.lists(st.integers()).map(sorted)
```
```python
>>> sorted_list_of_ints.example()
[-27120, 97, 12805]
```
<!-- #endregion -->

**Getting creative with the `.map` method: Solution**

Construct a Hypothesis strategy that produces either the string `"cat"` or the string `"dog"`.
Then, write a test that uses this strategy;
it should simply test that either `"cat"` or `"dog"` was indeed produced by the strategy.
Run the test

<!-- #region -->
```python
# We "hijack" the `st.booleans()` strategy, which only generates 
# `True` or `False`, and use the `.map` method to transform these
# two outputs to `"cat"` or `"dog"`.
#
# This is only one of many ways that you could have created this
# strategy

> cat_or_dog = st.booleans().map(lambda x: "cat" if x else "dog")

Reviewer 2: I'd prefer to show `st.sampled_from(["cat", "dog"])` as the primary example, with the remapped booleans as a curiosity.
